bin/cloud-provision.sh

```diff
@@ -16,6 +16,53 @@ fi
 NOVARC_PATH=$1
 SPI_CREDENTIALS_PATH=$2
 OPENSTACK_CREDENTIALS_DIR="$JENKINS_HOME/.openstack"
+PRIVATE_KEY_PATH=~/.canonistack/${OS_USERNAME}_${OS_REGION_NAME}.key
+
+wait_for_ssh(){
+    local ip="$1"
+    local id="$2"
+    retry=30
+    while ! ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no ubuntu@"$ip" true; do
+        retry=$(( retry - 1 ))
+        if [ $retry -le 0 ]; then
+            echo "Timed out waiting for ssh. Aborting!"
+            openstack server delete "$id"
+            exit 1
+        fi
+        sleep 10
+    done
+}
+
+wait_for_ip(){
+    local INSTANCE_ID=$1
+    local retry=30
+    local INSTANCE_IP=""
+    INSTANCE_IP=$(openstack server show $INSTANCE_ID | grep 'addresses' | awk '{print $4}' | cut -d= -f2)
+    # when the instance hasn't came up yet the addresses line reads:
+    #   | addresses |            |
+    # so print $4 would be '|', this value will be returned until we have an IP assigned
+    while [ -z "$INSTANCE_IP" -o "$INSTANCE_IP" = "|" ]; do
+        retry=$(( retry - 1 ))
+        if [ $retry -le 0 ]; then
+            echo "Timed out waiting for instance IP. Aborting!"
+            openstack server delete "$INSTANCE_ID"
+            exit 1
+        fi
+        sleep 20
+        INSTANCE_IP=$(openstack server show $INSTANCE_ID | grep 'addresses' | awk '{print $4}' | cut -d= -f2)
+    done
+    echo "$INSTANCE_IP"
+}
+
+launch_instance(){
+    local IMAGE_NAME=$1
+
+    IMAGE_ID=$(openstack image list | grep "$IMAGE_NAME" | head -1 | awk '{print $4}')
```

Comment:
In fact we don't need the image id, the name is enough to launch the instance. I've used the `-c` and `-f` options later to get the instance ip.